Examples/Algorithms/Geant4HepMC/src/EventRecording.cpp

```diff
@@ -0,0 +1,89 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "ActsExamples/EventRecording/EventRecording.hpp"
+#include "ActsExamples/EventData/SimParticle.hpp"
+#include "ActsExamples/Framework/WhiteBoard.hpp"
+#include "ActsExamples/Geant4/GdmlDetectorConstruction.hpp"
+#include <iostream>
+#include <stdexcept>
+#include <FTFP_BERT.hh>
+#include "EventAction.hpp"
+#include "G4RunManager.hh"
+#include "PrimaryGeneratorAction.hpp"
+#include "RunAction.hpp"
+#include "SteppingAction.hpp"
+
+#include <HepMC3/GenEvent.h>
+
+ActsExamples::EventRecording::~EventRecording() {
+  m_runManager = nullptr;
+}
+
+ActsExamples::EventRecording::EventRecording(
+    ActsExamples::EventRecording::Config&& cnf, Acts::Logging::Level level)
+    : ActsExamples::BareAlgorithm("EventRecording", level),
+      m_cfg(std::move(cnf)),
+      m_runManager(std::make_unique<G4RunManager>()) {
+  if (m_cfg.inputParticles.empty()) {
+    throw std::invalid_argument("Missing input particle collection");
+  }
+  if (m_cfg.outputHepMcTracks.empty()) {
+    throw std::invalid_argument("Missing output event collection");
+  }
+  if (!m_cfg.detectorConstruction) {
+    throw std::invalid_argument("Missing detector construction object");
+  }
+
+  /// Now set up the Geant4 simulation
+  m_runManager->SetUserInitialization(m_cfg.detectorConstruction.release());
+  m_runManager->SetUserInitialization(new FTFP_BERT);
+  m_runManager->SetUserAction(new ActsExamples::RunAction());
+  m_runManager->SetUserAction(new ActsExamples::EventAction());
+  m_runManager->SetUserAction(
+      new ActsExamples::PrimaryGeneratorAction(m_cfg.seed1, m_cfg.seed2));
+  m_runManager->SetUserAction(new ActsExamples::SteppingAction());
+  m_runManager->Initialize();
+}
+
+ActsExamples::ProcessCode ActsExamples::EventRecording::execute(
+    const ActsExamples::AlgorithmContext& context) const {
+  // ensure exclusive access to the geant run manager
+  std::lock_guard<std::mutex> guard(m_runManagerLock);
+
+  // Retrieve the initial particles
+  const auto initialParticles =
+      context.eventStore.get<ActsExamples::SimParticleContainer>(
+          m_cfg.inputParticles);
+
+  // Storage of events that will be produced
+  std::vector<HepMC3::GenEvent> events;
+  events.reserve(initialParticles.size());
+
+  for (const auto& part : initialParticles) {
+    // Prepare the particle gun
+    ActsExamples::PrimaryGeneratorAction::instance()->prepareParticleGun(part);
+
+    // Begin with the simulation
+    m_runManager->BeamOn(1);
+
+    HepMC3::GenEvent event = ActsExamples::EventAction::instance()->event();
+    HepMC3::FourVector shift(0., 0., 0., part.time());
+    event.shift_position_by(shift);
```

Comment:
See comment below regarding `SetParticleTime` for the particle gun. If this can be set through the particle gun directly, the explicit shift becomes unnecessary.

Reply:
If the supported lcg releases and the therewith supported G4 versions change one can consider this.

Reply:
Ok. Thanks for checking.